unittests/unit/test_PackageResourceRetriever.cpp

```diff
@@ -136,6 +141,116 @@ TEST(PackageResourceRetriever, exists_FallsBackOnSecondUri)
   ASSERT_EQ(2u, mockRetriever->mExists.size());
   EXPECT_EQ(expected1, mockRetriever->mExists[0]);
   EXPECT_EQ(expected2, mockRetriever->mExists[1]);
+  EXPECT_TRUE(mockRetriever->mGetFilePath.empty());
+  EXPECT_TRUE(mockRetriever->mRetrieve.empty());
+}
+
+TEST(PackageResourceRetriever, getFilePath_UnableToResolve_ReturnsEmptyString)
+{
+  auto mockRetriever = std::make_shared<PresentResourceRetriever>();
+  PackageResourceRetriever retriever(mockRetriever);
+
+  EXPECT_EQ(retriever.getFilePath(Uri::createFromString("package://test/foo")), "");
+  EXPECT_TRUE(mockRetriever->mExists.empty());
+  EXPECT_TRUE(mockRetriever->mGetFilePath.empty());
+  EXPECT_TRUE(mockRetriever->mRetrieve.empty());
+}
+
+TEST(PackageResourceRetriever, getFilePath_DelegateFails_ReturnsEmptyString)
+{
+  // GTest breaks the string concatenation.
+#ifdef _WIN32
+  const char* expected = "file:///" "dart://sample/test/foo";
+#else
+  const char* expected = "file://" "dart://sample/test/foo";
+#endif
+
+  auto mockRetriever = std::make_shared<AbsentResourceRetriever>();
+  PackageResourceRetriever retriever(mockRetriever);
+  retriever.addPackageDirectory("test", "dart://sample/test");
+
+  EXPECT_EQ(retriever.getFilePath(
+      Uri::createFromString("package://test/foo")), "");
+  ASSERT_TRUE(mockRetriever->mExists.empty());
+  EXPECT_EQ(expected, mockRetriever->mGetFilePath.front());
+  ASSERT_EQ(1u, mockRetriever->mGetFilePath.size());
+  EXPECT_TRUE(mockRetriever->mRetrieve.empty());
+}
+
+TEST(PackageResourceRetriever, getFilePath_UnsupportedUri_ReturnsEmptyString)
+{
+  auto mockRetriever = std::make_shared<PresentResourceRetriever>();
+  PackageResourceRetriever retriever(mockRetriever);
+  retriever.addPackageDirectory("test", "dart://sample/test");
+
+  EXPECT_EQ(retriever.getFilePath(Uri::createFromString("foo://test/foo")), "");
+  EXPECT_TRUE(mockRetriever->mExists.empty());
+  EXPECT_TRUE(mockRetriever->mGetFilePath.empty());
+  EXPECT_TRUE(mockRetriever->mRetrieve.empty());
+}
+
+TEST(PackageResourceRetriever, getFilePath_StripsTrailingSlash)
+{
+#ifdef _WIN32
+  const char* expected = "file:///" "dart://sample/test/foo";
```

Comment:
I'm very unclear about the purpose behind this `#ifdef`, especially since I believe this will set the value of `expected` to a `const char*` which dereferences to `file:///dart://sample/test/foo`, which doesn't seem like a valid URI (although I could certainly be wrong).

Some commentary about the motivation behind the `#ifdef` (Why is Windows different? Shouldn't that kind of difference be handled in the `Uri` class?) would be helpful. As well as the meaning behind strange URI which seems to contain both the `file` scheme and the `dart` scheme in sequence.

Reply:
Those invalid URI tests were unintentionally introduced in https://github.com/dartsim/dart/commit/162b7a842ee737c03339c6d6554954748fe69e46#diff-2c76e43b1bdbf70403566db01e2e1198. They are fixed by c79de2d.

> Some commentary about the motivation behind the #ifdef (Why is Windows different? Shouldn't that kind of difference be handled in the Uri class?) would be helpful.

Added. I hope the comment answers your concern.